### PR TITLE
Adding model-control-mode

### PR DESCRIPTION
**Pull request by udaij12**
## Description
Adding a model-control-mode to TorchServe #3158

Default is `none` which prevents users from registering and unregistering models after startup. 
Can be set to `enabled` which allows users to use register and unregister APIs

Check documentation for how to set the model-control-mode

## Tests
Setting model mode to "enabled" for all current tests. 
Added a new pytest [test_model_control_mode.py](https://github.com/pytorch/serve/pull/3165/files#diff-ca05115302bff77e9d174711f74dd947c611fc5ffb7c8fdf939bedf178637e33) that covers default and model api enabled

### CODE REVIEW

File: docs/model_control_mode.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
+# Model Control Mode
+
+TorchServe now supports model control mode with two settings "none"(default) and "explicit"
+
+## How to set Model Control to explicit
+1. Add `--model-mode-explicit` to command line when running TorchServe.
+2. Add `model_control_mode=explicit` to config.properties file
```
Comment (lxning) [inline docs/model_control_mode.md:7]: What's the priority b/w the two options? We can only offer this config.properties if it is too complicate for you to implement.

File: docs/model_control_mode.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    },
+    {
+      "modelName": "squeezenet1_1",
+      "modelUrl": "https://torchserve.pytorch.org/mar_files/squeezenet1_1.mar"
+    }
+  ]
+}
+ubuntu@ip-172-31-11-32:~/serve$ torchserve --stop
+TorchServe has stopped.
+```
```
Comment (lxning) [inline docs/model_control_mode.md:152]: it is not necessary to post the log in the readme. You only need to demo accepting register/unregister curl response at most.

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 3]
```diff
@@ -129,6 +130,11 @@ private boolean isManagementReq(String[] segments) {
                 || endpointMap.containsKey(segments[1]);
     }
 
+    private boolean checkModelControl() {
+        ConfigManager configManager = ConfigManager.getInstance();
+        return (configManager.getModelControlMode().equals("explicit"));
+    }
+
```
Comment (lxning) [inline frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:137]: 1. it is not efficient to compare string each time. 
2. better to rename as isModelControlled
3. move it to ConfigManager.
4. and also it is kind of redundancy getModelControlMode

File: docs/model_control_mode.md (hunk 1 above)
Comment (udaij12) [reply docs/model_control_mode.md:7]: Priority follows TorchServe standard as documented here https://github.com/pytorch/serve/blob/master/docs/configuration.md, will add a note in the docs that links to the priority

File: docs/model_control_mode.md (hunk 2 above)
Comment (udaij12) [reply docs/model_control_mode.md:152]: Will clean up the logs so it shows the register/unregister curl and output

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 3 above)
Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:137]: Noted will move the check into ConfigManager

**Review by lxning — COMMENTED** [review]
Can you add test for the priority b/w config file vs cmd, especially for the negative test cases (eg. conflict setting)?

File: frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java [hunk 4]
```diff
@@ -478,6 +484,11 @@ public int getNumberOfGpu() {
         return getIntProperty(TS_NUMBER_OF_GPU, 0);
     }
 
+    public boolean getModelControlMode() {
+        String mode = getProperty(MODEL_CONTROL_MODE, "none");
+        return mode.equals("explicit") ? true : false;
+    }
+
```
Comment (lxning) [inline frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:491]: As I commented before, it is not efficient to compare every time. it should be only time job, return true/false directly starting from 2nd time

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 3 above)
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java]: issue 1 is still not addressed.

File: frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java (hunk 4 above)
Comment (lxning) [reply frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:491]: I mean "mode.equals("explicit")" is not efficient.

Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:491]: It is only being compared one time now, like you mentioned that its not efficient every time. Are you saying to remove the string comparison all together? Currently it is only being compared once instead of during every check as before.

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 3 above)
Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java]: Issue 1 has been resolved as the string comparison only happens once and then is a true/false directly starting from 2nd time

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 5]
```diff
@@ -129,6 +132,14 @@ private boolean isManagementReq(String[] segments) {
...
     }
 
+    private boolean isModeExplicit() {
+        if (Modelmode == null) {
+            ConfigManager configManager = ConfigManager.getInstance();
+            Modelmode = configManager.getModelControlMode();
+        }
+        return Modelmode;
+    }
+
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:142]: This can be moved to the config manager `getModelControlMode`, so we can just query the model control mode here from the config manager and not have to store the configuration value.

File: test/pytest/test_model_control_mode.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    params = (
+        ("model_name", "resnet-18"),
+        ("url", "resnet-18.mar"),
+        ("initial_workers", "1"),
+        ("synchronous", "true"),
+    )
+    response = requests.post("http://localhost:8081/models", params=params)
+
+    assert response.status_code == 200, "model control check failed"
+    response = requests.get("http://localhost:8081/models/resnet-18")
```
Comment (namannandan) [inline test/pytest/test_model_control_mode.py:91]: Please extend test to include model deletion with `mode:explicit` to ensure we cover that use case as well.

**Review by namannandan — CHANGES_REQUESTED** [review]

File: docs/model_control_mode.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  "code": 405,
+  "type": "MethodNotAllowedException",
+  "message": "Requested method is not allowed, please refer to API document."
+}
+```
+
+## Model Control Mode EXPLICIT
+Setting model control to `explicit` allows users to load and unload models using the model load APIs.
+
+### Example using cmd line to set mode to explicit
```
[the remark this answers is not in the document]
Comment (udaij12) [reply docs/model_control_mode.md:43]: The default mode example already exists with no CLI input or config properties.

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 8]
```diff
@@ -48,6 +49,8 @@
  */
 public class ManagementRequestHandler extends HttpRequestHandlerChain {
 
+    private Boolean Modelmode = null;
```
[the remark this answers is not in the document]
Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java]: Noted and moved

File: test/pytest/test_model_control_mode.py (hunk 6 above)
Comment (udaij12) [reply test/pytest/test_model_control_mode.py:91]: Noted and added

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 9]
```diff
@@ -74,7 +75,7 @@ public void handleRequest(
                     if (HttpMethod.GET.equals(method)) {
                         handleListModels(ctx, decoder);
                         return;
-                    } else if (HttpMethod.POST.equals(method)) {
+                    } else if (HttpMethod.POST.equals(method) && isModeExplicit()) {
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:78]: We'll need to include this check for gRPC endpoints as well
https://github.com/pytorch/serve/blob/dd1f8dc36efe21f036be1e65043310dec74af28e/frontend/server/src/main/java/org/pytorch/serve/grpcimpl/ManagementImpl.java#L104

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 10]
```diff
@@ -93,7 +94,7 @@ public void handleRequest(
                     } else {
                         handleScaleModel(ctx, decoder, segments[2], modelVersion);
                     }
-                } else if (HttpMethod.DELETE.equals(method)) {
+                } else if (HttpMethod.DELETE.equals(method) && isModeExplicit()) {
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:97]: We'll need to include this check for gRPC as well
https://github.com/pytorch/serve/blob/dd1f8dc36efe21f036be1e65043310dec74af28e/frontend/server/src/main/java/org/pytorch/serve/grpcimpl/ManagementImpl.java#L196

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 11]
```diff
@@ -129,6 +130,11 @@ private boolean isManagementReq(String[] segments) {
                 || endpointMap.containsKey(segments[1]);
     }
 
+    private boolean isModeExplicit() {
+        ConfigManager configManager = ConfigManager.getInstance();
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:134]: Nit: `ConfigManager` is singleton so we can fetch this only once and save it in a class variable and just do `return configManager.getModelControlMode()` here.

File: frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java [hunk 12]
```diff
@@ -254,6 +256,11 @@ private ConfigManager(Arguments args) throws IOException {
             prop.setProperty(TS_LOAD_MODELS, String.join(",", models));
         }
 
+        boolean mode = args.isModeExplicit();
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:259]: Nit: 
```suggestion
        boolean modelControlMode = args.isModeControlModeExplicit();
```

File: frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java [hunk 13]
```diff
@@ -1117,6 +1134,7 @@ public static final class Arguments {
         private boolean snapshotDisabled;
         private String workflowStore;
         private String cppLogConfigFile;
+        private boolean modelExplicit;
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:1137]:
```suggestion
        private boolean modelControlModeExplicit;
```

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 9 above)
Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:78]: Created issue for gRPC

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 10 above)
Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:97]: Created issue for gRPC

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java [hunk 14]
```diff
@@ -48,9 +49,12 @@
  */
 public class ManagementRequestHandler extends HttpRequestHandlerChain {
 
+    ConfigManager configManager;
+
     /** Creates a new {@code ManagementRequestHandler} instance. */
     public ManagementRequestHandler(Map<String, ModelServerEndpoint> ep) {
         endpointMap = ep;
+        configManager = ConfigManager.getInstance();
```
Comment (namannandan) [inline frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:57]: Make member variable private
```suggestion
        private configManager = ConfigManager.getInstance();
```

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 9 above)
Comment (namannandan) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:81]: https://github.com/pytorch/serve/issues/3180

File: frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java (hunk 10 above)
Comment (namannandan) [reply frontend/server/src/main/java/org/pytorch/serve/http/api/rest/ManagementRequestHandler.java:100]: https://github.com/pytorch/serve/issues/3180

File: test/pytest/sanity/coverage.xml
Comment (namannandan) [inline test/pytest/sanity/coverage.xml:1]: Please remove this file from your PR.

**Review by namannandan — APPROVED** [review]
Few minor comments to address. Looks good otherwise!

File: docs/model_control_mode.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    Result: Explicit mode enabled
+* Example 2:
+  * Config file: `model_control_mode=explicit`
+
+    cmd line: `torchserve --start --ncs --model-store model_store`
+
+    Result: Mode is explicit (no way to enable "none" through cmd)
+
+## Model Control Mode Default
+At startup TorchServe loads only those models specified explicitly with the `--models` command-line option. After startup users will be unable to register or delete models in this mode.
```
Comment (mreso) [inline docs/model_control_mode.md:24]: The naming of the modes (none, explicit) does not make sense for TorchServe. Using the wording "explicitly" in the doc for the default (none) mode will confuse users even more.

The difference between other solutions (where this naming makes sense) is that TorchServe does not load all models in model_store by default but none. Only if we give --models we load the specified. So for TorchServe this PR is more about locking the api after startup so no models can be loaded afterwards. This should also be reflected in the parameter. 

To make clear what the parameter does we will need to name it something like: --model-control-api-enabled

File: frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java [hunk 16]
```diff
@@ -478,6 +485,14 @@ public int getNumberOfGpu() {
         return getIntProperty(TS_NUMBER_OF_GPU, 0);
     }
 
+    public boolean getModelControlMode() {
+        String mode = getProperty(MODEL_CONTROL_MODE, "none");
+        if (modelMode == null) {
```
Comment (mreso) [inline frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:490]: To avoid getting the property getProperty should be inside the condition.

File: test/pytest/test_utils.py [hunk 17]
```diff
@@ -59,6 +59,8 @@ def start_torchserve(
     no_config_snapshots=False,
     gen_mar=True,
     plugin_folder=None,
+    models=None,
+    mode=None,
```
Comment (mreso) [inline test/pytest/test_utils.py:63]: The default mode is called "none" in the docs. But here None selects explicit. Better to revert the logic to no confuse people.

**Review by mreso — CHANGES_REQUESTED** [review]
The chosen parameter name is not reflecting the actual functionality in TorchServe. Please see comments.

File: docs/model_control_mode.md (hunk 15 above)
Comment (udaij12) [reply docs/model_control_mode.md:24]: changed it to --model-api-enabled and changed explicit to enabled in docs and code

File: frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java (hunk 16 above)
Comment (udaij12) [reply frontend/server/src/main/java/org/pytorch/serve/util/ConfigManager.java:489]: done

File: test/pytest/test_utils.py (hunk 17 above)
Comment (udaij12) [reply test/pytest/test_utils.py:63]: The logic for test is to keep enabled as default so that it does not break existing test structure for now. Will require complete test overhaul at a later date.

File: docs/model_control_mode.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
+# Model Control Mode
+
+TorchServe now supports model control mode with two settings "none"(default) and "enabled"
```
Comment (namannandan) [inline docs/model_control_mode.md:3]: Please add a short description here of what the two modes do and which APIs they disable/enable.

File: docs/model_control_mode.md [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+## Two ways to set Model Control
+1. Add `--model-api-enabled` to command line when running TorchServe to switch from none to enabled mode. Command line cannot be used to set mode to none, can only be used to set to enabled
+2. Add `model_control_mode=enabled` or `model_control_mode=none` to config.properties file
```
Comment (namannandan) [inline docs/model_control_mode.md:7]: To simplify this configuration, please change it to `model_api_enabled` or `model_control_api_enabled` and accept boolean values.

File: docs/model_control_mode.md [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+Priority between cmd and config file follows the following [TorchServer standard](https://github.com/pytorch/serve/blob/c74a29e8144bc12b84196775076b0e8cf3c5a6fc/docs/configuration.md#advanced-configuration)
+* Example 1:
+  * Config file: `model_control_mode=none`
+
+    cmd line: `torchserve --start --ncs --model-store model_store --model-api-enabled`
+
+    Result: Model api mode enabled
```
Comment (namannandan) [inline docs/model_control_mode.md:15]: Describe here the values that were set in the config file and command line and which took precedence to make it clear to the reader.

File: docs/model_control_mode.md [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+* Example 2:
+  * Config file: `model_control_mode=enabled`
+
+    cmd line: `torchserve --start --ncs --model-store model_store`
+
+    Result: Mode is enabled (no way to enable "none" through cmd)
```
Comment (namannandan) [inline docs/model_control_mode.md:21]: Same as above: Describe here the values that were set in the config file and command line and which took precedence to make it clear to the reader.

File: test/pytest/test_utils.py (hunk 17 above)
Comment (mreso) [reply test/pytest/test_utils.py:63]: You can just set the default value to model the old behavior to not break any test.

Comment (udaij12) [reply test/pytest/test_utils.py:63]: We want TorchServe default to be model api enabled as false meaning that users will not be able to register or delete models once it starts. That change will break the tests

Comment (mreso) [reply test/pytest/test_utils.py]: Udai, this is not about the way people will setup TorchServe, this is only about the interface of start_torchserve which is not user facing.

File: test/pytest/test_model_control_mode.py [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+
+
+@pytest.fixture(scope="module")
+def setup_torchserve():
+    MODEL_STORE = os.path.join(ROOT_DIR, "model_store/")
+
+    Path(test_utils.MODEL_STORE).mkdir(parents=True, exist_ok=True)
+
+    test_utils.start_torchserve(
+        no_config_snapshots=True, models="mnist=mnist.mar", mode="none"
```
Comment (mreso) [inline test/pytest/test_model_control_mode.py]: The interface should be 
```
def start_torchserve(..., model_control_enabled=True):
```
so when you setup a test that want the new behavior you use:
```
test_utils.start_torchserve(..., model_control_enabled=False)
```

**Review by mreso — APPROVED** [review]
Addressed my own comments to move on with this. LGTM now
